### DISCUSSION TITLE
Wrong openapi response model with optional fields

### DISCUSSION THREAD

**Discussion by FArseny** [fastapi/fastapi#10041] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel
from fastapi import Request


class ResponseInfo(BaseModel):
    status: str | None = None
    entry_id: int


app = FastAPI()

@app.post("/info", response_model=ResponseInfo)
async def read_item(request: Request):
    return ResponseInfo(entry_id=10)
```

### Description

After running server i expect to see response model to have optional status, but swagger show it as required.
<img width="481" alt="Screenshot 2023-08-08 at 13 41 52" src="https://github.com/tiangolo/fastapi/assets/23155553/616f4370-c36c-4871-985f-100e6cf30672">

I checked `app.openapi()` it returns `... 'required': ['status', 'entry_id']`
And check pydantic:
`{'properties': {'status': {'anyOf': [{'type': 'string'}, {'type': 'null'}], 'default': None, 'title': 'Status'}, 'entry_id': {'title': 'Entry Id', 'type': 'integer'}}, 'required': ['entry_id'], 'title': 'ResponseInfo', 'type': 'object'}`

This takes place only in response_model; in request works as expected

### Operating System

macOS

### FastAPI Version

0.101.0

### Pydantic Version

2.1.1

### Python Version

3.11.0

**Comment by andrerav** [comment]
Seeing the same issue. This will cause problems with response deserialization if you generate a client from this spec in a language such as C#.  Pydantic 2.1.1, FastAPI 0.101.0.

**Reply by andrerav** [reply]
Can confirm that downgrading to 0.100.0 fixes the issue. I suppose #10011 is the culprit then :)

**Reply by andrerav** [reply]
@tiangolo FYI

**Reply by commonism** [reply]
related https://github.com/tiangolo/fastapi/discussions/10039

**Reply by alexgaspard** [reply]
We have the same issue and indeed we are using [openapi-generator-cli](https://github.com/OpenAPITools/openapi-generator-cli) to generate a client so this impacts us. Is there any workaround?

**Reply by andrerav** [reply]
@alexgaspard Downgrading to 0.100.0 worked well for me, try that and see if it helps :)

**Reply by alexgaspard** [reply]
@andrerav Thank you, yes it works! :)

**Comment by aravindnujella** [comment]
It generates `anyOf` with request models as well. 
https://github.com/tiangolo/fastapi/discussions/10070

**Comment by tiangolo** [comment, marked as answer]
Yep, there was a lot of work put to get this new behavior, it's actually more correct, but I just added a feature to allow disabling it if you want to get the old behavior back, it's available since FastAPI 0.102.0 🎉  https://fastapi.tiangolo.com/how-to/separate-openapi-schemas/

**Reply by aravindnujella** [reply]
@tiangolo Looks like fastapi (v0.103) still generates redundant anyOf schemas for optional fields. I have documented one case where redundant anyOf description is generated here - https://github.com/tiangolo/fastapi/discussions/10070
